src/Microsoft.VisualStudio.Threading.Analyzers.Tests/VSTHRD109AvoidAssertInAsyncMethodsAnalyzerTests.cs

```diff
@@ -112,6 +227,43 @@ public async Task TaskReturningLambdaInsideVoidMethod_GeneratesDiagnostic()
 using Microsoft.VisualStudio.Shell;
 using Task = System.Threading.Tasks.Task;
 
+class Test {
+    void Foo() {
+        Task.Run<int>(() => {
+            ThreadHelper.ThrowIfNotOnUIThread();
+            return Task.FromResult(1);
+        });
+    }
+}
+";
+
+            var fix = @"
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Shell;
+using Task = System.Threading.Tasks.Task;
+
+class Test {
+    void Foo() {
+        Task.Run<int>(async () => {
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+            return 1;
+        });
+    }
+}
+";
+
+            var expected = Verify.Diagnostic().WithSpan(9, 26, 9, 46);
+            await Verify.VerifyCodeFixAsync(test, expected, fix);
+        }
+
+        [Fact(Skip = "Fails because the semantic model claims the anonymous func returns Task instead of Task<int>. We should probably skip that check and always preserve return statements.")]
```

Comment:
📝 This may work in newer versions of Roslyn with later language versions (bestest betterness). If the underlying issue is the inability to distinguish between `Task.Run(Func<Task>)` and `Task.Run<TResult>(Func<Task<TResult>>)`, you can force a disambiguation in the test by calling `Task.Run<int>` instead of just `Task.Run`.

Reply:
Yes, using `Task.Run<int>` does resolve the issue (see the test directly above this one, on which I applied that workaround). 

Interestingly though, although this compiles:

```cs
Task<int> t = Task.Run(() => Task.FromResult(1));
```

Even in that case, which compiles, Roslyn *still* claims that the delegate returns `Task`. Yet if it really did, the result couldn't be assigned to `Task<int>` as I did. So what's up with that? Would 'bestest betterness' actually solve this?

Reply:
> Even in that case, which compiles,

Are you compiling with the same version of Roslyn as you are using to run the tests?

Reply:
No. I compile the analyzers against Roslyn 1.2, and the tests run with 2.3. Evidently the dependency on the new Roslyn test framework bumped up the version of Roslyn I'm testing with.
How does that affect this investigation?

Reply:
> Evidently the dependency on the new Roslyn test framework bumped up the version of Roslyn I'm testing with.

This shouldn't be the case. The test framework is compiled against Roslyn 1.0.1. The version of Roslyn used for testing is controlled by the following line:

https://github.com/Microsoft/vs-threading/blob/9df4cbf729555494c4a3b2ef9902fbb7dfa81100/src/Microsoft.VisualStudio.Threading.Analyzers.Tests/Microsoft.VisualStudio.Threading.Analyzers.Tests.csproj#L23

Git blame indicates this was updated to Roslyn 2.3.2 in #190.